uarray/_uarray_dispatch.cxx

```diff
@@ -0,0 +1,479 @@
+	return (value == def);
+}
+
+
+py_ref Function::canonicalize_args(PyObject * args)
+{
+	const auto arg_size = PyTuple_GET_SIZE(args);
+	const auto def_size = PyTuple_GET_SIZE(def_args_.get());
+
+	if (arg_size > def_size)
+		return py_ref::ref(args);
+
+	Py_ssize_t mismatch = 0;
+	for (Py_ssize_t i = arg_size - 1; i >= 0; --i)
+	{
+		auto val = PyTuple_GET_ITEM(args, i);
+		auto def = PyTuple_GET_ITEM(def_args_.get(), i);
+		if (!is_default(val, def))
+		{
+			mismatch = i + 1;
+			break;
+		}
+	}
+
+	return py_ref::steal(PyTuple_GetSlice(args, 0, mismatch));
+}
+
+
+py_ref Function::canonicalize_kwargs(PyObject * kwargs)
+{
+	if (kwargs == nullptr)
+		return py_ref::steal(PyDict_New());
+
+	PyObject * key, * def_value;
+	Py_ssize_t pos = 0;
+	while (PyDict_Next(def_kwargs_, &pos, &key, &def_value))
+	{
+		auto val = PyDict_GetItem(kwargs, key);
+		if (is_default(val, def_value))
+		{
+			PyDict_DelItem(kwargs, key);
+		}
+	}
+	return py_ref::ref(kwargs);
+}
+
+
+py_func_args Function::replace_dispatchables(
+	PyObject * backend, PyObject * args, PyObject * kwargs, PyObject * coerce)
+{
+	auto ua_convert =
+		py_ref::steal(PyObject_GetAttrString(backend, "__ua_convert__"));
+
+	if (!ua_convert)
+	{
+		PyErr_Clear();
+		return {py_ref::ref(args), py_ref::ref(kwargs)};
+	}
+
+	auto dispatchables = py_ref::steal(PyObject_Call(extractor_, args, kwargs));
+	if (!dispatchables)
+		return {};
+
+	auto convert_args = py_make_tuple(dispatchables, coerce);
+	auto res = py_ref::steal(PyObject_Call(ua_convert, convert_args, nullptr));
+	if (!res)
+	{
+		return {};
+	}
+
+	if (res == Py_NotImplemented)
+	{
+		return {std::move(res), nullptr};
+	}
+
+	auto itr = py_ref::steal(PyObject_GetIter(res));
+	if (itr == nullptr)
+	{
+		return {};
+	}
+
+	const auto num_disp = PyTuple_Size(dispatchables);
+	auto replaced_args = py_ref::steal(PyTuple_New(num_disp));
+	Py_ssize_t pos = 0;
+	while (auto item = py_ref::steal(PyIter_Next(itr)))
+	{
+		if (pos > num_disp)
+		{
+			PyErr_SetString(PyExc_ValueError,
+			                "Too many values returned from __ua_convert__");
+			return {};
+		}
+		PyTuple_SET_ITEM(replaced_args.get(), pos, item.release());
+		++pos;
+	}
+
+	auto replacer_args = py_make_tuple(args, kwargs, replaced_args);
+	if (!replacer_args)
+		return {};
+
+	res = py_ref::steal(PyObject_Call(replacer_, replacer_args, nullptr));
+	if (PyTuple_Size(res) != 2)
+		return {};
+
+	auto new_args = py_ref::ref(PyTuple_GET_ITEM(res.get(), 0));
+	auto new_kwargs = py_ref::ref(PyTuple_GET_ITEM(res.get(), 1));
+
+	new_kwargs = canonicalize_kwargs(new_kwargs);
+
+	if (!PyTuple_Check(new_args) || !PyDict_Check(new_kwargs))
+	{
+		PyErr_SetString(PyExc_ValueError, "Invalid return from argument_replacer");
+		return {};
+	}
+
+	return {std::move(new_args), std::move(new_kwargs)};
+}
+
+
+PyObject * uarray_function_call(
+	PyObject * self, PyObject * args, PyObject * kwargs)
+{
+	return reinterpret_cast<Function *>(self)->call(args, kwargs);
```

Comment:
Here, I believe you need to inject `obj` into the `args` if it isn't `nullptr`.

Reply:
The `PyMethod` wrapper already takes care of that:
```
>>> class Foo: pass
>>> Foo.f = func
>>> Foo().f
<bound method extractor of <__main__.Foo object at 0x7f0f9d128630>>
```